### PR TITLE
Upgrade handling for IM4.0

### PR DESCRIPTION
**Pull request by ujjwalchk-it**
This PR includes two changes -

1. Removing list of old CP2 deployments while installing CP3 
```
"auth-idp", "auth-pdp", "auth-pap", "secret-watcher", "oidcclient-watcher", "iam-policy-controller"
```

2. Enhanced create/update scenario for `ibmcloud-cluster-info` and secret `ibmcloud-cluster-ca-cert` while `management-ingress` still getting uninstalled. IM operator now look for `ownerReference` and `label` to determine if the running artefacts  owned by`management-ingress`. Based on that it wait for the reconcile and proceed with reconcile once those are removed (`management-ingress`) is removed.

#57512
#57516

### CODE REVIEW

File: pkg/controller/authentication/deployment.go [hunk 1]
```diff
@@ -37,6 +37,29 @@ func (r *ReconcileAuthentication) handleDeployment(instance *operatorv1alpha1.Au
...
+	cp2deployment := [6]string{"auth-idp", "auth-pdp", "auth-pap", "secret-watcher", "oidcclient-watcher", "iam-policy-controller"}
+
+	// Check for existing CP2 Deployments , Delete those if found
+	for i := 0; i < len(cp2deployment); i++ {
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: cp2deployment[i], Namespace: instance.Namespace}, currentDeployment)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				//reqLogger.Info("Upgrade check : deployment is not found.", "Deployment.Namespace", instance.Namespace, "Deployment.Name", cp2deployment[i])
+			} else {
+				reqLogger.Info("Upgrade check : Error while getting depoyment.", "Deployment.Namespace", instance.Namespace, "Deployment.Name", cp2deployment[i], "Error.Message", err)
```
Comment (rwhundley) [inline pkg/controller/authentication/deployment.go:50]:
```suggestion
				reqLogger.Info("Upgrade check : Error while getting deployment.", "Deployment.Namespace", instance.Namespace, "Deployment.Name", cp2deployment[i], "Error.Message", err)
```

File: pkg/controller/authentication/deployment.go [hunk 2]
```diff
@@ -37,6 +37,29 @@ func (r *ReconcileAuthentication) handleDeployment(instance *operatorv1alpha1.Au
...
 	reqLogger := log.WithValues("Instance.Namespace", instance.Namespace, "Instance.Name", instance.Name)
 
+	// We need to cleanup existing cp2 deployment before the CP3 installation
+	cp2deployment := [6]string{"auth-idp", "auth-pdp", "auth-pap", "secret-watcher", "oidcclient-watcher", "iam-policy-controller"}
+
+	// Check for existing CP2 Deployments , Delete those if found
+	for i := 0; i < len(cp2deployment); i++ {
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: cp2deployment[i], Namespace: instance.Namespace}, currentDeployment)
+		if err != nil {
+			if errors.IsNotFound(err) {
```
Comment (rwhundley) [inline pkg/controller/authentication/deployment.go:47]: Should convert this to !errors.IsNotFound(err) and remove the commented out line assuming the log line isn't wanted.

**Review by rwhundley — CHANGES_REQUESTED** [review]

File: pkg/controller/authentication/deployment.go (hunk 2 above)
Comment (ujjwalchk-it) [reply pkg/controller/authentication/deployment.go]: Done.

File: pkg/controller/authentication/deployment.go (hunk 1 above)
Comment (ujjwalchk-it) [reply pkg/controller/authentication/deployment.go]: Spelling corrected.

**Review by rashmi43 — APPROVED** [review]

**Pull request merged**